### PR TITLE
 Decompile viewport management

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

File: src/openloco/gui.cpp [hunk 1]
```diff
@@ -36,7 +37,7 @@ namespace openloco::gui
             (ui::window_event_list*)0x004FA1F4);
         window->widgets = _mainWindowWidgets;
         addr<0x00e3f0b8, int32_t>() = 0; // gCurrentRotation?
-        openloco::ui::viewportmgr::create(window, window->x, window->y, window->width, window->height, false, 0, 6143, 6143, 480);
+        openloco::ui::viewportmgr::create(window, 0, { window->x, window->y }, { window->width, window->height }, viewportmgr::ZoomLevel::full, { (map::map_rows * map::tile_size) / 2 - 1, (map::map_rows * map::tile_size) / 2 - 1, 480 });
```
Comment (AaronVanGeffen) [inline src/openloco/gui.cpp]: For clarity, I think these coordinate arguments deserve a line of their own.

**Comment by marijnvdwerf** [comment]
Can we possibly get rid of [C4458](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-4-c4458?view=vs-2017) (<q>declaration of '*identifier*' hides class member</q>). It's quite annoying to work around.

**Comment by AaronVanGeffen** [comment]
Hmm, I'm not sure we should. Shadowing can come back to bite you unexpectedly, so I think these warnings serve their purpose. I can see why it's cumbersome to fix these afterwards, though... 

Personally, I'd go further and try enabling the same errors on GCC/Clang as well. Would help in automatically setting up IDEs so they warn while coding, too, I suppose.

**Comment by janisozaur** [comment]
I'm with @AaronVanGeffen on that one

File: src/openloco/viewportmgr.h [hunk 2]
```diff
@@ -1,13 +1,29 @@
...
 
 namespace openloco::ui::viewportmgr
 {
-    constexpr size_t max_viewports = 10;
+    enum class ZoomLevel : uint8_t
+    {
+        full = 0,
+        half = 1,
+        quarter = 2,
+        eight = 3,
```
Comment (AaronVanGeffen) [inline src/openloco/viewportmgr.h]:
```suggestion
        eighth = 3,
```

File: src/openloco/station.h [hunk 3]
```diff
@@ -50,7 +50,11 @@ namespace openloco
     struct station
     {
         string_id name; // 0x00
-        uint8_t pad_02[0x28 - 0x02];
+        uint8_t pad_02[0x08 - 0x02];
+        uint16_t var_08[4];
```
Comment (AaronVanGeffen) [inline src/openloco/station.h]:
```suggestion
        uint16_t labels_left[4];
```

etc

File: src/openloco/viewportmgr.cpp [hunk 4]
```diff
@@ -268,6 +263,15 @@ namespace openloco::ui::viewportmgr
         }
     }
 
+    /**
+     * 0x004CBB01 (eight)
```
Comment (AaronVanGeffen) [inline src/openloco/viewportmgr.cpp]: `eighth` ;)

File: src/openloco/viewportmgr.cpp [hunk 5]
```diff
@@ -217,14 +224,26 @@ namespace openloco::ui::viewportmgr
...
+        if (doGarbageCollect)
+        {
+            collectGarbage();
+        }
     }
 
     // 0x004CBA2D
     void invalidate(station* station)
     {
+        bool doGarbageCollect;
```
Comment (AaronVanGeffen) [inline src/openloco/viewportmgr.cpp]: Initialise to zero.

**Pull request merged**